src/app/clusters/operational-state-server/operational-state-delegate.h

```diff
@@ -0,0 +1,393 @@
+    {
+        completionErrorCode = aCompletionErrorCode;
+        totalOperationalTime.SetValue(chip::app::DataModel::Nullable<uint32_t>(aTotalOperationalTime));
+        pausedTime.SetValue(chip::app::DataModel::Nullable<uint32_t>(aPausedTime));
+    }
+};
+
+/**
+ * A delegate to handle application logic of the Operational State aliased Cluster.
+ */
+class Delegate
+{
+public:
+    /**
+     * Get operational state.
+     * @param void.
+     * @return the const reference of operational state.
+     */
+    virtual const GenericOperationalState & GetOperationalState() const = 0;
+
+    /**
+     * Get operational state list.
+     * @param operationalStateList The pointer to operational state list.
+     * After a successful return the caller is responsible for calling ReleaseOperationalStateList on the outparam.
+     */
+    virtual CHIP_ERROR GetOperationalStateList(GenericOperationalStateList ** operationalStateList, size_t & size) = 0;
+
+    /**
+     * Release operational state list
+     * @param operationalStateList The pointer for which to clear the OperationalStateStructDynamicList.
+     * @return void
+     */
+    virtual void ReleaseOperationalStateList(GenericOperationalStateList * operationalStateList) = 0;
+
+    /**
+     * Get operational phase list.
+     * @param operationalPhaseList The pointer to operational phase list.
+     * After a successful return the caller is responsible for calling ReleaseOperationalPhaseList on the outparam.
+     */
+    virtual CHIP_ERROR GetOperationalPhaseList(GenericOperationalPhaseList ** operationalPhaseList, size_t & size) = 0;
+
+    /**
+     * Release operational phase list
+     * @param operationalStateList The pointer for which to clear the GenericOperationalPhaseList.
+     * @return void
+     */
+    virtual void ReleaseOperationalPhaseList(GenericOperationalPhaseList * operationalPhaseList) = 0;
+
+    /**
+     * Get operational error.
+     * @param void.
+     * @return the const reference of operational error.
+     */
+    virtual const GenericOperationalError GetOperationalError() const = 0;
+
+    /**
+     * Set operational state.
+     * @param opState The operational state for which to set.
+     * @return CHIP_ERROR CHIP_NO_ERROR on success, or corresponding error code.
+     */
+    virtual CHIP_ERROR SetOperationalState(const GenericOperationalState & opState) = 0;
+
+    /**
+     * Set operational error.
+     * @param opErrState The reference of operational error.
+     * @return CHIP_ERROR CHIP_NO_ERROR on success, or corresponding error code.
+     */
+    virtual CHIP_ERROR SetOperationalError(const GenericOperationalError & opErrState) = 0;
+
+    // command callback
+    /**
+     * Handle Command Callback in application: Pause
+     * @param[out] get operational error after callback.
+     * @return operational error after callback
+     */
+    virtual GenericOperationalError & HandlePauseStateCallback(GenericOperationalError & err) = 0;
```

Comment:
Why do we have the same thing as both a return value and an outparam?  This applies to all the functions.